### PR TITLE
Iprep feature 6857/v1

### PR DESCRIPTION
**Pull request by victorjulien**
Implementation of `isset` and `isnotset` for the `iprep` keyword.

https://redmine.openinfosecfoundation.org/issues/6857

Looking for:
- code review
- logic feedback
- testing

### CODE REVIEW

**Comment by suricata-qa** [comment]
Information: QA ran without warnings. 

Pipeline 20591

File: rust/src/detect/iprep.rs [hunk 1]
```diff
@@ -75,36 +76,46 @@ extern "C" {
...
-    let (i, _) = char(',')(i)?;
-    let (i, _) = opt(is_a(" "))(i)?;
+pub fn detect_parse_iprep(i: &str) -> IResult<&str, DetectIPRepData, RuleParseError<&str>> {
+    // Inner utility function for easy error creation.
+    fn make_error(reason: String) -> nom7::Err<RuleParseError<&'static str>> {
+        Err::Error(RuleParseError::InvalidIPRep(reason))
+    }
+    let (_, values) = nom7::multi::separated_list1(
+        tag(","),
+        preceded(multispace0, nom7::bytes::complete::is_not(",")),
```
Comment (catenacyber) [inline rust/src/detect/iprep.rs:86]: Why this change ?

This is normally used if you do not care about the subparts of a keyword

File: rust/src/detect/iprep.rs [hunk 2]
```diff
@@ -75,36 +76,46 @@ extern "C" {
...
+    )(i)?;
+
+    if values.len() == 4 {
+        let cmd = match DetectIPRepDataCmd::from_str(values[0].trim()) {
```
Comment (catenacyber) [inline rust/src/detect/iprep.rs:90]: If let Ok(cmd) = else return Err

File: rust/src/detect/iprep.rs [hunk 3]
```diff
@@ -112,6 +112,28 @@ pub fn detect_parse_iprep(i: &str) -> IResult<&str, DetectIPRepData, RuleParseEr
...
+        };
+        let name = values[1].trim();
+        let namez = CString::new(name).unwrap();
+        let cat = unsafe { SRepCatGetByShortname(namez.as_ptr()) };
+        if cat == 0 {
+            return Err(make_error("unknown category".to_string()));
+        }
+        let (mode, arg1) = match values[2].trim() {
+            "isset" => { (DetectUintMode::DetectUintModeGte, 0)}, // any value means it's set
+            "isnotset" => { (DetectUintMode::DetectUintModeEqual, 255) }, // impossible value means not set
```
[the remark this answers is not in the document]
Comment (catenacyber) [reply rust/src/detect/iprep.rs:128]: Another solution would be to create `DetectUintModeAny` and `DetectUintModeNone` ?

Or to have the wrapper structure `DetectIPRepData` use booleans `justest` and `isnotset`

File: rust/src/detect/iprep.rs (hunk 1 above)
Comment (victorjulien) [reply rust/src/detect/iprep.rs:86]: Guess I'm not really sure what other ways there are to do it. I just want to split by `,` or end of string.

Comment (victorjulien) [reply rust/src/detect/iprep.rs:86]: btw took this directly from the bytemath parsing code, even though that has a slightly different layout

File: rust/src/detect/iprep.rs (hunk 2 above)
Comment (victorjulien) [reply rust/src/detect/iprep.rs:90]: do you mean something like this?
```diff
-    if values.len() == 4 {
-        let cmd = match DetectIPRepDataCmd::from_str(values[0].trim()) {
-            Ok(val) => val,
-            Err(_) => return Err(make_error("invalid command".to_string())),
+    let args = values.len();
+    if args == 4 || args == 3 {
+        let cmd = if let Ok(cmd) = DetectIPRepDataCmd::from_str(values[0].trim()) {
+            cmd
+        } else {
+            return Err(make_error("invalid command".to_string()));
         };
```
Gets more verbose, not sure I see this at better.

File: rust/src/detect/iprep.rs [hunk 4]
```diff
@@ -75,36 +76,46 @@ extern "C" {
...
+            Err(_) => return Err(make_error("invalid value".to_string())),
+        };
+        let du8 = DetectUintData::<u8> {
+            arg1,
+            arg2: 0,
+            mode,
+        };
+        return Ok((i, DetectIPRepData { du8, cat, cmd }));
+    } else {
+        return Err(make_error("too many arguments".to_string()));
```
[the remark this answers is not in the document]
Comment (victorjulien) [reply rust/src/detect/iprep.rs:138]: ack

File: rust/src/detect/iprep.rs [hunk 5]
```diff
@@ -112,6 +112,28 @@ pub fn detect_parse_iprep(i: &str) -> IResult<&str, DetectIPRepData, RuleParseEr
             mode,
         };
         return Ok((i, DetectIPRepData { du8, cat, cmd }));
+    } else if values.len() == 3 {
+        let cmd = match DetectIPRepDataCmd::from_str(values[0].trim()) {
```
[the remark this answers is not in the document]
Comment (victorjulien) [reply rust/src/detect/iprep.rs:116]: ack, deduplicating

File: rust/src/detect/iprep.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (victorjulien) [reply rust/src/detect/iprep.rs:128]: `DetectUintModeGt` 255 is an invalid setting for u8, even if it is perhaps not enforced anywhere.

Would like 255 to explicit, or even use a named constant to make it easy to understand what it signifies.

[the remark this answers is not in the document]
Comment (victorjulien) [reply rust/src/detect/iprep.rs:128]: I guess `DetectUintModeAny` could make some sense, though it could just be conceptualized as `true`, but less sure about `DetectUintModeNone`. I think your suggestion of not overloading the uint support is probably the cleanest.

**Comment by victorjulien** [comment]
replaced by #11091

File: rust/src/detect/iprep.rs (hunk 1 above)
Comment (catenacyber) [reply rust/src/detect/iprep.rs:86]: > btw took this directly from the bytemath parsing code

bytemath is right using this, because bytemath is a list of sub keywords like `endian` or `relative`, whose order is not relevant. bytemath parses kind of a list of key+values.

Here, you want to parse something different, you parse a list of specific things...

> I just want to split by , or end of string.

I think it is easier to have some linear code like currently, but yes, if you really want to first split into words/fields, this is good.

But then, you should not duplicate code, cf below...

File: rust/src/detect/iprep.rs (hunk 2 above)
Comment (catenacyber) [reply rust/src/detect/iprep.rs:90]: Would something like `let cmd =  DetectIPRepDataCmd::from_str(values[0].trim())?` work ?

Current code is `    let (i, cmd) = map_res(alpha0, DetectIPRepDataCmd::from_str)(i)?;`

So, maybe something like  `et (_, cmd) = map_res(alpha0, DetectIPRepDataCmd::from_str)(values[0].trim())?;`

File: rust/src/detect/iprep.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (catenacyber) [reply rust/src/detect/iprep.rs:128]: > have the wrapper structure DetectIPRepData use booleans justest and isnotset

Looks cleanest indeed
